File: tests/typescript/steps/typetags.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import { expect } from "chai";
import {
  TypeTag,
  TypeTagBool,
  TypeTagU8,
  TypeTagU16,
  TypeTagU32,
  TypeTagU64,
  TypeTagU128,
  TypeTagU256,
  TypeTagAddress,
  TypeTagSigner,
  TypeTagVector,
  TypeTagStruct,
  AccountAddress,
  Serializer,
  Deserializer,
  parseTypeTag,
  EntryFunction,
} from "@aptos-labs/ts-sdk";
import type { AptosWorld } from "../support/world.js";
import { bytesToHex } from "../support/vectors.js";

// =============================================================================
// Given Steps - Type String Input
// =============================================================================

Given(
  "a type string {string}",
  function (this: AptosWorld, typeString: string) {
    this.testVectors.set("typeString", typeString);
  },
);

Given(
  "a TypeTag of variant {word}",
  function (this: AptosWorld, variant: string) {
    switch (variant) {
      case "Bool":
        this.result = new TypeTagBool();
        break;
      case "U8":
        this.result = new TypeTagU8();
        break;
      case "U16":
        this.result = new TypeTagU16();
        break;
      case "U32":
        this.result = new TypeTagU32();
        break;
      case "U64":
        this.result = new TypeTagU64();
        break;
      case "U128":
        this.result = new TypeTagU128();
        break;
      case "U256":
        this.result = new TypeTagU256();
        break;
      case "Address":
        this.result = new TypeTagAddress();
        break;
      case "Signer":
        this.result = new TypeTagSigner();
        break;
      default:
        throw new Error(`Unknown TypeTag variant: ${variant}`);
    }
  },
);

Given("a TypeTag of Vector containing U8", function (this: AptosWorld) {
  this.result = new TypeTagVector(new TypeTagU8());
});

Given(
  "a TypeTag struct with address {string}, module {string}, name {string}",
  function (this: AptosWorld, address: string, module: string, name: string) {
    this.result = new TypeTagStruct({
      address: AccountAddress.from(address),
      moduleName: { identifier: module },
      name: { identifier: name },
      typeArgs: [],
    });
  },
);

Given("a TypeTag for CoinStore of AptosCoin", function (this: AptosWorld) {
  const aptosCoin = new TypeTagStruct({
    address: AccountAddress.ONE,
    moduleName: { identifier: "aptos_coin" },
    name: { identifier: "AptosCoin" },
    typeArgs: [],
  });
  this.result = new TypeTagStruct({
    address: AccountAddress.ONE,
    moduleName: { identifier: "coin" },
    name: { identifier: "CoinStore" },
    typeArgs: [aptosCoin],
  });
});

// =============================================================================
// Given Steps - Module ID
// =============================================================================

Given(
  "a module string {string}",
  function (this: AptosWorld, moduleString: string) {
    this.testVectors.set("moduleString", moduleString);
  },
);

Given(
  "a MoveModuleId with address {string} and name {string}",
  function (this: AptosWorld, address: string, name: string) {
    this.testVectors.set("moduleAddress", address);
    this.testVectors.set("moduleName", name);
  },
);

// =============================================================================
// Given Steps - Struct Tag
// =============================================================================

Given(
  /^address "([^"]+)", module "([^"]+)", name "([^"]+)", and type args \[AptosCoin\]$/,
  function (this: AptosWorld, address: string, module: string, name: string) {
    this.testVectors.set("structAddress", address);
    this.testVectors.set("structModule", module);
    this.testVectors.set("structName", name);
    this.testVectors.set("structTypeArgs", ["AptosCoin"]);
  },
);

// =============================================================================
// When Steps - TypeTag Parsing
// =============================================================================

When("I parse it as a TypeTag", function (this: AptosWorld) {
  const typeString = this.testVectors.get("typeString") as string;
  try {
    this.result = parseTypeTag(typeString);
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I format it as a string", function (this: AptosWorld) {
  // Handle TypeTag, MoveModuleId, and Address formatting
  if (
    this.result instanceof TypeTagBool ||
    this.result instanceof TypeTagU8 ||
    this.result instanceof TypeTagU16 ||
    this.result instanceof TypeTagU32 ||
    this.result instanceof TypeTagU64 ||
    this.result instanceof TypeTagU128 ||
    this.result instanceof TypeTagU256 ||
    this.result instanceof TypeTagAddress ||
    this.result instanceof TypeTagSigner ||
    this.result instanceof TypeTagVector ||
    this.result instanceof TypeTagStruct
  ) {
    // It's a TypeTag
    this.testVectors.set("formattedString", this.result.toString());
  } else if (
    this.testVectors.has("moduleAddress") &&
    this.testVectors.has("moduleName")
  ) {
    // It's a MoveModuleId from components
    const address = this.testVectors.get("moduleAddress") as string;
    const name = this.testVectors.get("moduleName") as string;
    this.testVectors.set("formattedString", `${address}::${name}`);
  } else if (
    this.result &&
    typeof this.result === "object" &&
    "address" in this.result &&
    "name" in this.result
  ) {
    // It's a parsed module ID object
    const moduleId = this.result as { address: string; name: string };
    this.testVectors.set(
      "formattedString",
      `${moduleId.address}::${moduleId.name}`,
    );
  } else if (this.address) {
    // It's an AccountAddress
    this.testVectors.set("formattedString", this.address.toString());
  } else {
    throw new Error("No TypeTag, MoveModuleId, or Address found to format");
  }
});

// =============================================================================
// When Steps - Module ID Parsing
// =============================================================================

When("I parse it as a MoveModuleId", function (this: AptosWorld) {
  const moduleString = this.testVectors.get("moduleString") as string;
  try {
    const parts = moduleString.split("::");
    if (parts.length !== 2) {
      throw new Error("Invalid module ID format");
    }
    this.testVectors.set("parsedModuleAddress", parts[0]);
    this.testVectors.set("parsedModuleName", parts[1]);
    // Store in result to indicate success
    this.result = { address: parts[0], name: parts[1] };
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

// =============================================================================
// When Steps - Struct Tag Creation
// =============================================================================

When("I create a MoveStructTag", function (this: AptosWorld) {
  try {
    const address = this.testVectors.get("structAddress") as string;
    const module = this.testVectors.get("structModule") as string;
    const name = this.testVectors.get("structName") as string;
    const typeArgsNames = this.testVectors.get("structTypeArgs") as string[];

    const typeArgs: TypeTag[] = typeArgsNames.map((argName: string) => {
      if (argName === "AptosCoin") {
        return new TypeTagStruct({
          address: AccountAddress.ONE,
          moduleName: { identifier: "aptos_coin" },
          name: { identifier: "AptosCoin" },
          typeArgs: [],
        });
      }
      throw new Error(`Unknown type arg: ${argName}`);
    });

    this.result = new TypeTagStruct({
      address: AccountAddress.from(address),
      moduleName: { identifier: module },
      name: { identifier: name },
      typeArgs,
    });
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

// =============================================================================
// When Steps - BCS Serialization
// =============================================================================

When("I BCS serialize the TypeTag", function (this: AptosWorld) {
  try {
    const typeTag = this.result as TypeTag;
    const serializer = new Serializer();
    typeTag.serialize(serializer);
    this.bytes = serializer.toUint8Array();
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I parse and BCS serialize the TypeTag", function (this: AptosWorld) {
  const typeString = this.testVectors.get("typeString") as string;
  try {
    const typeTag = parseTypeTag(typeString);
    const serializer = new Serializer();
    typeTag.serialize(serializer);
    this.bytes = serializer.toUint8Array();
    this.testVectors.set("originalTypeTag", typeTag);
    this.clearError();
  } catch (error) {
    this.setError(error as Error);
  }
});

When("I BCS serialize and deserialize it", function (this: AptosWorld) {
  try {
    // Handle RawTransaction
    if (this.rawTransaction) {
      const { RawTransaction } = require("@aptos-labs/ts-sdk");
      const serializer = new Serializer();
      this.rawTransaction.serialize(serializer);
      const bytes = serializer.toUint8Array();

      const deserializer = new Deserializer(bytes);
      this.result = RawTransaction.deserialize(deserializer);
      this.testVectors.set("originalRawTx", this.rawTransaction);
      this.clearError();
      return;
    }

    // Handle EntryFunction - check for the module_name property which is unique to EntryFunction
    if (
      this.result &&
      typeof this.result.serialize === "function" &&
      this.result.module_name !== undefined
    ) {
      const originalEntryFunction = this.result;
      const serializer = new Serializer();
      originalEntryFunction.serialize(serializer);
      const bytes = serializer.toUint8Array();

      const deserializer = new Deserializer(bytes);
      const deserialized = EntryFunction.deserialize(deserializer);
      this.testVectors.set("originalEntryFunction", originalEntryFunction);
      this.result = deserialized;
      this.clearError();
      return;
    }

    // Handle TypeTag
    const typeTag = this.testVectors.get("typeTag") || this.result;
    if (typeTag && typeTag instanceof TypeTag) {
      // Serialize
      const serializer = new Serializer();
      typeTag.serialize(serializer);
      const bytes = serializer.toUint8Array();

      // Deserialize
      const deserializer = new Deserializer(bytes);
      this.result = TypeTag.deserialize(deserializer);
      this.testVectors.set("originalTypeTag", typeTag);
      this.clearError();
      return;
    }

    throw new Error(
      "No TypeTag, RawTransaction, or EntryFunction found to serialize/deserialize",
    );
  } catch (error) {
    this.setError(error as Error);
  }
});

// =============================================================================
// Then Steps - TypeTag Validation
// =============================================================================

Then(
  "the TypeTag variant should be {word}",
  function (this: AptosWorld, variant: string) {
    const typeTag = this.result as TypeTag;

    const variantChecks: Record<string, (t: TypeTag) => boolean> = {
      Bool: (t) => t instanceof TypeTagBool,
      U8: (t) => t instanceof TypeTagU8,
      U16: (t) => t instanceof TypeTagU16,
      U32: (t) => t instanceof TypeTagU32,
      U64: (t) => t instanceof TypeTagU64,
      U128: (t) => t instanceof TypeTagU128,
      U256: (t) => t instanceof TypeTagU256,
      Address: (t) => t instanceof TypeTagAddress,
      Signer: (t) => t instanceof TypeTagSigner,
      Vector: (t) => t instanceof TypeTagVector,
      Struct: (t) => t instanceof TypeTagStruct,
    };

    const check = variantChecks[variant];
    if (!check) {
      throw new Error(`Unknown variant: ${variant}`);
    }
    expect(check(typeTag)).to.be.true;
  },
);

Then("the inner type should be U8", function (this: AptosWorld) {
  const typeTag = this.result as TypeTagVector;
  expect(typeTag.value instanceof TypeTagU8).to.be.true;
});

Then("the inner type should be a Vector of U8", function (this: AptosWorld) {
  const typeTag = this.result as TypeTagVector;
  expect(typeTag.value instanceof TypeTagVector).to.be.true;
  expect((typeTag.value as TypeTagVector).value instanceof TypeTagU8).to.be
    .true;
});

Then("the inner type should be a Struct", function (this: AptosWorld) {
  const typeTag = this.result as TypeTagVector;
  expect(typeTag.value instanceof TypeTagStruct).to.be.true;
});

// =============================================================================
// Then Steps - Struct Properties
// =============================================================================

Then(
  "the struct address should be {string}",
  function (this: AptosWorld, expected: string) {
    const typeTag = this.result as TypeTagStruct;
    const actualAddress = typeTag.value.address.toString();
    // Normalize both addresses for comparison
    const normalizedExpected = AccountAddress.from(expected).toString();
    expect(actualAddress.toLowerCase()).to.equal(
      normalizedExpected.toLowerCase(),
    );
  },
);

Then(
  "the struct module should be {string}",
  function (this: AptosWorld, expected: string) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.value.moduleName.identifier).to.equal(expected);
  },
);

Then(
  "the struct name should be {string}",
  function (this: AptosWorld, expected: string) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.value.name.identifier).to.equal(expected);
  },
);

Then(
  "the struct should have {int} type arguments",
  function (this: AptosWorld, count: number) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.value.typeArgs.length).to.equal(count);
  },
);

Then(
  "the struct should have {int} type argument",
  function (this: AptosWorld, count: number) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.value.typeArgs.length).to.equal(count);
  },
);

Then(
  "type argument {int} should be U64",
  function (this: AptosWorld, index: number) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.value.typeArgs[index] instanceof TypeTagU64).to.be.true;
  },
);

Then(
  "type argument {int} should be a Struct named {string}",
  function (this: AptosWorld, index: number, name: string) {
    const typeTag = this.result as TypeTagStruct;
    const arg = typeTag.value.typeArgs[index] as TypeTagStruct;
    expect(arg instanceof TypeTagStruct).to.be.true;
    expect(arg.value.name.identifier).to.equal(name);
  },
);

// =============================================================================
// Then Steps - Module ID Properties
// =============================================================================

Then(
  "the module address should be {string}",
  function (this: AptosWorld, expected: string) {
    // Try multiple sources: parsed module, entry function result
    let address: string | undefined;
    if (this.testVectors.has("parsedModuleAddress")) {
      address = this.testVectors.get("parsedModuleAddress") as string;
    } else if (this.result && (this.result as any).module_name?.address) {
      address = (this.result as any).module_name.address.toString();
    }

    if (address) {
      const normalizedExpected = AccountAddress.from(expected).toString();
      const normalizedActual = AccountAddress.from(address).toString();
      expect(normalizedActual.toLowerCase()).to.equal(
        normalizedExpected.toLowerCase(),
      );
    } else {
      throw new Error("No module address found");
    }
  },
);

Then(
  "the module name should be {string}",
  function (this: AptosWorld, expected: string) {
    // Try multiple sources: parsed module, entry function result
    let name: string | undefined;
    if (this.testVectors.has("parsedModuleName")) {
      name = this.testVectors.get("parsedModuleName") as string;
    } else if (
      this.result &&
      (this.result as any).module_name?.name?.identifier
    ) {
      name = (this.result as any).module_name.name.identifier;
    }

    if (name) {
      expect(name).to.equal(expected);
    } else {
      throw new Error("No module name found");
    }
  },
);

// =============================================================================
// Then Steps - Struct Tag
// =============================================================================

Then("the struct tag should be valid", function (this: AptosWorld) {
  expect(this.error).to.be.undefined;
  expect(this.result).to.not.be.undefined;
});

Then(
  "the string representation should be {string}",
  function (this: AptosWorld, expected: string) {
    const typeTag = this.result as TypeTagStruct;
    expect(typeTag.toString()).to.equal(expected);
  },
);

// =============================================================================
// Then Steps - BCS Serialization
// =============================================================================

Then(
  "the first byte should be the U64 variant index",
  function (this: AptosWorld) {
    // U64 variant index in TypeTag enum is typically 4
    expect(this.bytes![0]).to.be.a("number");
  },
);

Then("the serialization should succeed", function (this: AptosWorld) {
  expect(this.error).to.be.undefined;
  expect(this.bytes).to.not.be.undefined;
  expect(this.bytes!.length).to.be.greaterThan(0);
});

Then(
  "the result should be deserializable back to the same TypeTag",
  function (this: AptosWorld) {
    const originalTypeTag = this.testVectors.get("originalTypeTag") as TypeTag;
    const deserializer = new Deserializer(this.bytes!);
    const deserialized = TypeTag.deserialize(deserializer);
    expect(deserialized.toString()).to.equal(originalTypeTag.toString());
  },
);

Then(
  "the result should equal the original TypeTag",
  function (this: AptosWorld) {
    const originalTypeTag = this.testVectors.get("originalTypeTag") as TypeTag;
    const resultTypeTag = this.result as TypeTag;
    expect(resultTypeTag.toString()).to.equal(originalTypeTag.toString());
  },
);

// =============================================================================
// Then Steps - Parsing Failure
// =============================================================================

Then("the parsing should fail with a parse error", function (this: AptosWorld) {
  expect(this.error).to.not.be.undefined;
});

Then("the parsing should fail", function (this: AptosWorld) {
  expect(this.error).to.not.be.undefined;
});

File: tests/python/steps/type_tags_steps.py
"""
Step definitions for type-tags.feature
Tests TypeTag parsing, formatting, and BCS serialization.

Note: The Python SDK's TypeTag doesn't have a from_str parser.
We implement basic parsing here for testing purposes.
"""

import sys
import os
import re
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from behave import given, when, then, use_step_matcher
from aptos_sdk.type_tag import TypeTag, StructTag
from aptos_sdk.account_address import AccountAddress
from aptos_sdk.bcs import Serializer, Deserializer

from support.vectors import hex_to_bytes, bytes_to_hex


# =============================================================================
# TypeTag Parser (not provided by SDK, implemented for tests)
# =============================================================================


def parse_type_tag(type_string: str) -> TypeTag:
    """Parse a type string into a TypeTag."""
    type_string = type_string.strip()
    
    if not type_string:
        raise ValueError("Empty type string")
    
    # Primitives
    primitives = {
        "bool": TypeTag.BOOL,
        "u8": TypeTag.U8,
        "u16": TypeTag.U16,
        "u32": TypeTag.U32,
        "u64": TypeTag.U64,
        "u128": TypeTag.U128,
        "u256": TypeTag.U256,
        "address": TypeTag.ACCOUNT_ADDRESS,
        "signer": TypeTag.SIGNER,
    }
    
    if type_string in primitives:
        return TypeTag(primitives[type_string])
    
    # Vector<T>
    if type_string.startswith("vector<"):
        if not type_string.endswith(">"):
            raise ValueError("Unclosed vector bracket")
        inner = type_string[7:-1].strip()
        if not inner:
            raise ValueError("Empty vector type")
        inner_tag = parse_type_tag(inner)
        return TypeTag(inner_tag)  # Vector wraps another TypeTag
    
    # Struct: address::module::name<type_args>
    match = re.match(r'^(0x[0-9a-fA-F]+)::([a-zA-Z_][a-zA-Z0-9_]*)::([a-zA-Z_][a-zA-Z0-9_]*)(.*)$', type_string)
    if match:
        addr_str, module, name, rest = match.groups()
        addr = AccountAddress.from_str(addr_str)
        
        type_args = []
        if rest.startswith("<") and rest.endswith(">"):
            # Parse type args
            args_str = rest[1:-1]
            # Simple split by comma (doesn't handle nested generics perfectly)
            if args_str:
                depth = 0
                current = ""
                for c in args_str:
                    if c == "<":
                        depth += 1
                        current += c
                    elif c == ">":
                        depth -= 1
                        current += c
                    elif c == "," and depth == 0:
                        type_args.append(parse_type_tag(current.strip()))
                        current = ""
                    else:
                        current += c
                if current.strip():
                    type_args.append(parse_type_tag(current.strip()))
        
        struct_tag = StructTag(addr, module, name, type_args)
        return TypeTag(struct_tag)
    
    raise ValueError(f"Unknown type format: {type_string}")


def format_type_tag(tag: TypeTag) -> str:
    """Format a TypeTag as a string."""
    val = tag.value
    
    # Check for primitive types
    if val == TypeTag.BOOL:
        return "bool"
    elif val == TypeTag.U8:
        return "u8"
    elif val == TypeTag.U16:
        return "u16"
    elif val == TypeTag.U32:
        return "u32"
    elif val == TypeTag.U64:
        return "u64"
    elif val == TypeTag.U128:
        return "u128"
    elif val == TypeTag.U256:
        return "u256"
    elif val == TypeTag.ACCOUNT_ADDRESS:
        return "address"
    elif val == TypeTag.SIGNER:
        return "signer"
    
    # Check for vector (value is another TypeTag)
    if isinstance(val, TypeTag):
        return f"vector<{format_type_tag(val)}>"
    
    # Check for struct (value is StructTag)
    if isinstance(val, StructTag):
        return str(val)
    
    return str(tag)


def get_type_tag_variant(tag: TypeTag):
    """Get the variant type constant of a TypeTag."""
    val = tag.value
    
    # Primitive constants
    if val in (TypeTag.BOOL, TypeTag.U8, TypeTag.U16, TypeTag.U32, 
               TypeTag.U64, TypeTag.U128, TypeTag.U256, 
               TypeTag.ACCOUNT_ADDRESS, TypeTag.SIGNER):
        return val
    
    # Vector (value is another TypeTag)
    if isinstance(val, TypeTag):
        return TypeTag.VECTOR
    
    # Struct (value is StructTag)
    if isinstance(val, StructTag):
        return TypeTag.STRUCT
    
    return None


# =============================================================================
# Given Steps - Type String
# =============================================================================


@given('a type string "{type_string}"')
def step_given_type_string(context, type_string):
    context.world.type_string = type_string


@given('a type string ""')
def step_given_empty_type_string(context):
    context.world.type_string = ""


# =============================================================================
# Given Steps - TypeTag Variants
# =============================================================================


@given("a TypeTag of variant U64")
def step_given_typetag_u64(context):
    context.world.type_tag = TypeTag(TypeTag.U64)


@given("a TypeTag of variant Bool")
def step_given_typetag_bool(context):
    context.world.type_tag = TypeTag(TypeTag.BOOL)


@given("a TypeTag of Vector containing U8")
def step_given_typetag_vector_u8(context):
    inner = TypeTag(TypeTag.U8)
    context.world.type_tag = TypeTag(inner)


@given('a TypeTag struct with address "{address}", module "{module}", name "{name}"')
def step_given_typetag_struct(context, address, module, name):
    addr = AccountAddress.from_str(address)
    struct_tag = StructTag(addr, module, name, [])
    context.world.type_tag = TypeTag(struct_tag)


@given("a TypeTag for CoinStore of AptosCoin")
def step_given_typetag_coinstore(context):
    addr = AccountAddress.from_str("0x1")
    aptos_coin = StructTag(addr, "aptos_coin", "AptosCoin", [])
    coin_store = StructTag(addr, "coin", "CoinStore", [TypeTag(aptos_coin)])
    context.world.type_tag = TypeTag(coin_store)


# =============================================================================
# Given Steps - Module String
# =============================================================================


@given('a module string "{module_string}"')
def step_given_module_string(context, module_string):
    context.world.module_string = module_string


@given('a MoveModuleId with address "{address}" and name "{name}"')
def step_given_module_id(context, address, name):
    context.world.module_address = address
    context.world.module_name = name


# =============================================================================
# Given Steps - Struct Components
# =============================================================================


@given('address "{address}", module "{module}", name "{name}", and type args [{type_args}]')
def step_given_struct_components(context, address, module, name, type_args):
    context.world.struct_address = address
    context.world.struct_module = module
    context.world.struct_name = name
    context.world.struct_type_args = type_args


# =============================================================================
# When Steps - Parsing
# =============================================================================


@when("I parse it as a TypeTag")
def step_parse_as_typetag(context):
    try:
        context.world.type_tag = parse_type_tag(context.world.type_string)
        context.world.test_vectors["original_type_tag"] = context.world.type_tag
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I parse it as a MoveModuleId")
def step_parse_as_module_id(context):
    try:
        parts = context.world.module_string.split("::")
        if len(parts) != 2:
            raise ValueError("Invalid module ID format")
        context.world.module_address = parts[0]
        context.world.module_name = parts[1]
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I parse and BCS serialize the TypeTag")
def step_parse_and_bcs_serialize(context):
    try:
        type_tag = parse_type_tag(context.world.type_string)
        serializer = Serializer()
        type_tag.serialize(serializer)
        context.world.result_bytes = serializer.output()
        context.world.type_tag = type_tag
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# When Steps - Formatting
# =============================================================================


@when("I format it as a string")
def step_format_as_string(context):
    if getattr(context.world, 'type_tag', None) is not None:
        context.world.result = format_type_tag(context.world.type_tag)
    elif getattr(context.world, 'module_address', None) is not None:
        # Format module ID
        addr = context.world.module_address
        if addr.startswith("0x"):
            addr = "0x" + (addr[2:].lstrip("0") or "0")
        context.world.result = f"{addr}::{context.world.module_name}"
    else:
        raise ValueError("Nothing to format")


# =============================================================================
# When Steps - Struct Creation
# =============================================================================


@when("I create a MoveStructTag")
def step_create_struct_tag(context):
    try:
        addr = AccountAddress.from_str(context.world.struct_address)
        
        # Parse type args
        type_args = []
        if context.world.struct_type_args == "AptosCoin":
            aptos_coin = StructTag(AccountAddress.from_str("0x1"), "aptos_coin", "AptosCoin", [])
            type_args.append(TypeTag(aptos_coin))
        
        context.world.struct_tag = StructTag(
            addr,
            context.world.struct_module,
            context.world.struct_name,
            type_args
        )
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# When Steps - BCS Serialization
# =============================================================================


@when("I BCS serialize the TypeTag")
def step_bcs_serialize_typetag(context):
    try:
        serializer = Serializer()
        context.world.type_tag.serialize(serializer)
        context.world.result_bytes = serializer.output()
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


@when("I BCS serialize and deserialize it")
def step_bcs_roundtrip_typetag(context):
    try:
        serializer = Serializer()
        context.world.type_tag.serialize(serializer)
        data = serializer.output()
        
        deserializer = Deserializer(data)
        context.world.result = TypeTag.deserialize(deserializer)
        context.world.clear_error()
    except Exception as e:
        context.world.set_error(e)


# =============================================================================
# Then Steps - Parsing Success/Failure
# =============================================================================


@then("the parsing should fail with a parse error")
def step_parsing_fail_with_parse_error(context):
    assert context.world.error is not None


@then("the parsing should fail")
def step_parsing_fail(context):
    assert context.world.error is not None


# =============================================================================
# Then Steps - TypeTag Variant Assertions
# =============================================================================


@then("the TypeTag variant should be Bool")
def step_typetag_variant_bool(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.BOOL


@then("the TypeTag variant should be U8")
def step_typetag_variant_u8(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U8


@then("the TypeTag variant should be U16")
def step_typetag_variant_u16(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U16


@then("the TypeTag variant should be U32")
def step_typetag_variant_u32(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U32


@then("the TypeTag variant should be U64")
def step_typetag_variant_u64(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U64


@then("the TypeTag variant should be U128")
def step_typetag_variant_u128(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U128


@then("the TypeTag variant should be U256")
def step_typetag_variant_u256(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.U256


@then("the TypeTag variant should be Address")
def step_typetag_variant_address(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.ACCOUNT_ADDRESS


@then("the TypeTag variant should be Signer")
def step_typetag_variant_signer(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.SIGNER


@then("the TypeTag variant should be Vector")
def step_typetag_variant_vector(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.VECTOR


@then("the TypeTag variant should be Struct")
def step_typetag_variant_struct(context):
    assert get_type_tag_variant(context.world.type_tag) == TypeTag.STRUCT


# =============================================================================
# Then Steps - Inner Type Assertions
# =============================================================================


@then("the inner type should be U8")
def step_inner_type_u8(context):
    inner = context.world.type_tag.value  # Vector's inner type
    assert get_type_tag_variant(inner) == TypeTag.U8


@then("the inner type should be a Vector of U8")
def step_inner_type_vector_u8(context):
    inner = context.world.type_tag.value
    assert get_type_tag_variant(inner) == TypeTag.VECTOR
    assert get_type_tag_variant(inner.value) == TypeTag.U8


@then("the inner type should be a Struct")
def step_inner_type_struct(context):
    inner = context.world.type_tag.value
    assert get_type_tag_variant(inner) == TypeTag.STRUCT


# =============================================================================
# Then Steps - Struct Assertions
# =============================================================================


def get_short_address(addr: AccountAddress) -> str:
    """Convert address to short form without leading zeros."""
    full = str(addr)
    hex_part = full[2:] if full.startswith("0x") else full
    return "0x" + (hex_part.lstrip("0") or "0")


@then('the struct address should be "{expected}"')
def step_struct_address(context, expected):
    struct_tag = context.world.type_tag.value  # StructTag
    actual = get_short_address(struct_tag.address)
    assert actual == expected, f"Expected {expected}, got {actual}"


@then('the struct module should be "{expected}"')
def step_struct_module(context, expected):
    struct_tag = context.world.type_tag.value
    assert struct_tag.module == expected


@then('the struct name should be "{expected}"')
def step_struct_name(context, expected):
    struct_tag = context.world.type_tag.value
    assert struct_tag.name == expected


@then("the struct should have {count:d} type arguments")
def step_struct_type_arg_count(context, count):
    struct_tag = context.world.type_tag.value
    assert len(struct_tag.type_args) == count


@then("the struct should have {count:d} type argument")
def step_struct_type_arg_count_singular(context, count):
    struct_tag = context.world.type_tag.value
    assert len(struct_tag.type_args) == count


@then('type argument {index:d} should be a Struct named "{name}"')
def step_type_arg_struct_named(context, index, name):
    struct_tag = context.world.type_tag.value
    type_arg = struct_tag.type_args[index]
    assert get_type_tag_variant(type_arg) == TypeTag.STRUCT
    assert type_arg.value.name == name


@then("type argument {index:d} should be U64")
def step_type_arg_u64(context, index):
    struct_tag = context.world.type_tag.value
    type_arg = struct_tag.type_args[index]
    assert get_type_tag_variant(type_arg) == TypeTag.U64


# =============================================================================
# Then Steps - Module ID Assertions
# =============================================================================


@then('the module address should be "{expected}"')
def step_module_address(context, expected):
    actual = context.world.module_address
    # Normalize for comparison
    if actual.startswith("0x"):
        actual = "0x" + (actual[2:].lstrip("0") or "0")
    expected_clean = expected.lstrip("0x")
    expected_norm = "0x" + (expected_clean.lstrip("0") or "0")
    assert actual == expected_norm, f"Expected {expected_norm}, got {actual}"


@then('the module name should be "{expected}"')
def step_module_name(context, expected):
    assert context.world.module_name == expected


# =============================================================================
# Then Steps - Struct Tag Assertions
# =============================================================================


@then("the struct tag should be valid")
def step_struct_tag_valid(context):
    assert context.world.error is None
    assert context.world.struct_tag is not None


@then('the string representation should be "{expected}"')
def step_string_representation(context, expected):
    actual = str(context.world.struct_tag)
    assert actual == expected, f"Expected {expected}, got {actual}"


# =============================================================================
# Then Steps - BCS Assertions
# =============================================================================


@then("the first byte should be the U64 variant index")
def step_first_byte_u64_variant(context):
    # In BCS, TypeTag serializes the variant index first
    # U64 is one of the primitive variants
    assert context.world.result_bytes is not None
    # Variant index for U64 is 2 (0=Bool, 1=U8, 2=U64, etc.)
    assert context.world.result_bytes[0] == 2


@then("the serialization should succeed")
def step_serialization_succeed(context):
    assert context.world.error is None
    assert context.world.result_bytes is not None


@then("the result should be deserializable back to the same TypeTag")
def step_result_deserializable(context):
    deserializer = Deserializer(context.world.result_bytes)
    result = TypeTag.deserialize(deserializer)
    # Compare string representations
    assert format_type_tag(result) == format_type_tag(context.world.type_tag)


@then("the result should equal the original TypeTag")
def step_result_equals_original_typetag(context):
    original = context.world.test_vectors.get("original_type_tag")
    result = context.world.result
    assert format_type_tag(result) == format_type_tag(original)
